src/guardian/keys-validation/keys-validation.service.ts

```diff
@@ -27,29 +43,63 @@ export class KeysValidationService {
   ): Promise<{ key: string; depositSignature: string }[]> {
     const forkVersion: Uint8Array = await this.forkVersion();
 
-    const validatedKeys: [
-      {
-        key: string;
-        depositSignature: string;
-        used: boolean;
-      },
-      boolean,
-    ][] = await this.keyValidator.validateKeys(
-      vettedKeys.map((key) => ({
-        key: key.key,
-        depositSignature: key.depositSignature,
-        used: false,
-        withdrawalCredentials: bufferFromHexString(withdrawalCredentials),
-        genesisForkVersion: Buffer.from(forkVersion.buffer),
-      })),
-    );
+    const keysForValidation = vettedKeys
+      .map((key) =>
+        this.filterKeyForValidation(key, withdrawalCredentials, forkVersion),
+      )
+      // filter keys that were in cache or signature wasn't changed
+      .filter((key) => key !== undefined) as DepositData[];
+
+    const validatedKeys: [Key & DepositData, boolean][] =
+      await this.keyValidator.validateKeys(keysForValidation);
+
+    this.updateCacheWithValidationResults(validatedKeys);
 
-    return validatedKeys
-      .filter(([, result]) => !result)
+    // this list will not include invalid keys from cache
+    const invalidKeysFromCurrentValidation = validatedKeys
+      .filter(([, isValid]) => !isValid)
       .map(([key]) => ({
         key: key.key,
         depositSignature: key.depositSignature,
       }));
+
+    // merge just checked invalid keys and invalid keys from cache but only from vettedKeys
+    return this.mergeInvalidKeys(vettedKeys, invalidKeysFromCurrentValidation);
+  }
+
+  filterKeyForValidation(
+    key: RegistryKey,
+    withdrawalCredentials: string,
+    forkVersion: Uint8Array,
+  ): DepositData | null {
+    const cachedEntry = this.keysCache.get(key.key);
+
+    // key was in cache and signature wasn't changed
+    if (this.keyNeedToBeValidated(cachedEntry, key)) {
+      return null;
+    }
+
+    return this.depositData(key, withdrawalCredentials, forkVersion);
+  }
+
+  keyNeedToBeValidated(
+    cachedEntry: { signature: string; isValid: boolean } | undefined,
+    key: RegistryKey,
+  ): boolean {
+    return !!cachedEntry && cachedEntry.signature == key.depositSignature;
```

Comment:
You have to check all key object, not only signature.

Bad scenario
User uploads key.
We validate that key
He deletes that key and uploads with same signature and pubkey, but with wrong creds or other field.
Signature is broken, but we think it's ok, because we already validated it